### DISCUSSION TITLE
How to build app resources with a particular directory layout?

### DISCUSSION THREAD

**Discussion by smilediver** [mesonbuild/meson#11725] [Q&A]
Hi. My question is how to use meson to build resources for an app, where the app is expecting a particular layout for the resources? Essentially I need to control where things go inside the build directory. Example of my use case is as follows:

Source data:
```
data/
  objects/
    house/
      house.png
      house.mesh
  shaders/
    shader.frag
```
Built data and what the app might expect:
```
data/ - (anywhere in build dir)
  objects/
    house/
      house.png.bin
      house.mesh
  shaders/
    glsl/
      shader.frag.bin
    dx12/
      shader.frag.bin
```
The build process basically has 3 cases:

1) Copy resource as is, while mirroring the layout
    `data/objects/house/house.mesh` -> `data/objects/house/house.mesh`
2) Compile resource, while mirroring the layout
    `data/objects/house/house.png` -> `data/objects/house/house.png.bin`
3) Compile resource and copy to other location than it actually is (note that destination might be some other directory, not necessarily a subdirectory of the source)
    `data/shaders/shader.frag` -> `data/shaders/glsl/shader.frag.bin`
    `data/shaders/shader.frag` -> `data/shaders/dx12/shader.frag.bin`

So I need some advice of what is the best way to set this up. Right now it seems that I need to mirror the build data directory in my source tree, place `meson.build` files where I want the built resources to go (eg `data/shaders/glsl/meson.build`), and use `custom_target()`. But this greatly complicates the whole build setup, because I can't have just a single `meson.build` file for data with a list of all resources and how to build them.

Any tips or examples would be appreciated.

**Comment by eli-schwartz** [comment]
Maybe you could use `meson install` and just have that produce the correct layout for runtime? This does sound more like an install-time thing to do.
